### PR TITLE
Added some features for ChaCha

### PR DESCRIPTION
**Pull request by SimCog**
# ChaCha additional features

## ChaCha20/8 8-round stream cipher

According to [1], the designer of ChaCha20 included 3 kinds of round ciphers:
- ChaCha20/20 (The original one)
- ChaCha20/12 (12-round stream cipher)
- ChaCha20/8 (8-round stream cipher)

Since there were 20 and 12, I added 8 and test it using the original implementation of Daniel J. Bernstein for getting the ciphertext: https://cr.yp.to/chacha.html

## Getter to the keystream generator

According to [2], more accurately the chapter 6 entitled _Stream Ciphers_, it is specified in the section **Synchronous stream ciphers** a function for generating the set of keystreams used by XORing it with the plaintext for creating the ciphertext. So, I implemented the method and added a m_initial_state. Thus, it is possible to generate a specific length keystream for a specific offset into the keystream.

# References

1. Bernstein, Daniel J. "ChaCha, a variant of Salsa20." In Workshop Record of SASC, vol. 8. 2008.
2. Menezes, Alfred J., Paul C. Van Oorschot, and Scott A. Vanstone. Handbook of applied cryptography. CRC press, 1996.

### CODE REVIEW

File: src/scripts/ci/travis/build.sh [hunk 1]
```diff
@@ -13,7 +13,8 @@ elif [ "$BUILD_MODE" = "sanitizer" ]; then
 fi
 
 if [ "$MODULES" = "min" ]; then
-    CFG_FLAGS+=(--minimized-build --enable-modules=base)
+    # Added chacha for passing the minimum modules compilation
+    CFG_FLAGS+=(--minimized-build --enable-modules=base,chacha)
```
Comment (randombit) [inline src/scripts/ci/travis/build.sh]: This change should not be required. Code in the tests should check if BOTAN_HAS_CHACHA is defined.

File: src/lib/stream/chacha/chacha.cpp [hunk 2]
```diff
@@ -173,4 +176,70 @@ std::string ChaCha::name() const
    return "ChaCha(" + std::to_string(m_rounds) + ")";
    }
 
+secure_vector<byte> ChaCha::get_keystream(u64bit offset, size_t length)
```
Comment (randombit) [inline src/lib/stream/chacha/chacha.cpp]: This API doesn't seem right to me. It would be better to add `StreamCipher::seek(u64bit byte_offset)` which moves the counter (for ChaCha, by manipulating m_state[12,13] and m_position to move to the byte offset within the keystream blocks - you already have the core of the logic in this function). Then encryption of all zeros produces the desired result. The same `seek` API could be implemented by CTR or Salsa.

Implementing something like `get_keystream` then becomes just

```
secure_vector<uint8_t> keystream(length);
cipher->seek(offset);
cipher->encipher(keystream);
return keystream;
```

The way you have it here, you can concurrently call `get_keystream` at high offsets and encrypt normally with the from-zero offsets. But that doesn't seem like a practical advantage, and doing it this way requires keeping a second copy of the state just in order to support this API. Having a single (per-object) byte position is simpler and seems to support the use case just as well.

**Comment by SimCog** [comment]
Thank you for your constructive remarks. I took them in considerations for this one. For the moment I did not implement the others stream ciphers.

File: src/lib/stream/chacha/chacha.cpp [hunk 3]
```diff
@@ -173,4 +172,27 @@ std::string ChaCha::name() const
...
    }
 
+void ChaCha::seek(u64bit offset)
+   {
+   if (m_state.size() == 0 && m_buffer.size() == 0)
+      {
+         throw Invalid_State("You have to setup the stream cipher (key and iv)");
+      }
+
+   m_position = offset;
```
Comment (randombit) [inline src/lib/stream/chacha/chacha.cpp]: m_position is the read index into the buffer - should `offset` be reduced modulo the block size here?

File: src/tests/data/stream/chacha.vec [hunk 4]
```diff
@@ -1,3 +1,9 @@
+[ChaCha(8)]
```
Comment (randombit) [inline src/tests/data/stream/chacha.vec]: Tests for the `seek` function are also needed. Add support in the tests (src/tests/test_stream.cpp) for an Seek parameter (which defaults to 0 for compatibility with existing tests), then using another implementation create ChaCha test vectors using at least Seek 1...65 and then also very large offsets.

**Comment by SimCog** [comment]
Thank you for the feedback. I rewrite the implementation as well as added bunch of tests.

File: src/lib/stream/chacha/chacha.h [hunk 5]
```diff
@@ -42,13 +42,15 @@ class BOTAN_DLL ChaCha final : public StreamCipher
...
+      void seek(u64bit offset) override;
+
    private:
       void key_schedule(const byte key[], size_t key_len) override;
 
       size_t m_rounds;
       secure_vector<u32bit> m_state;
       secure_vector<byte> m_buffer;
-      size_t m_position = 0;
+      u64bit m_position = 0;
```
Comment (randombit) [inline src/lib/stream/chacha/chacha.h]: Why this change? m_position is just an index into the (small) vector `m_buffer`, and could even (besides consistency concerns) just be a short rather than a size_t.

File: src/tests/test_stream.cpp [hunk 6]
```diff
@@ -53,6 +54,8 @@ class Stream_Cipher_Tests : public Text_Based_Test
             if(nonce.size())
                cipher->set_iv(nonce.data(), nonce.size());
 
+            cipher->seek(seek);
+
```
Comment (randombit) [inline src/tests/test_stream.cpp]: The test should only call if seek is set, since for example RC4 and OFB do not have a fast seek operation and will always throw.

File: src/lib/prov/openssl/openssl_rc4.cpp [hunk 7]
```diff
@@ -45,6 +45,11 @@ class OpenSSL_RC4 : public StreamCipher
 
       explicit OpenSSL_RC4(size_t skip = 0) : m_skip(skip) { clear(); }
       ~OpenSSL_RC4() { clear(); }
+
+      void seek(u64bit offset) override
+      {
+        (void)offset;
```
Comment (randombit) [inline src/lib/prov/openssl/openssl_rc4.cpp]: I understand why you did this, but this should throw an Exception instead. In the case of RC4, calling `seek` is an application logic error, so this particular case should be `throw Exception("RC4 does not support seeking")`

File: src/lib/stream/ctr/ctr.cpp [hunk 8]
```diff
@@ -125,4 +125,9 @@ void CTR_BE::increment_counter()
    m_pad_pos = 0;
    }
 
+void CTR_BE::seek(u64bit offset)
+   {
+      (void)offset;
```
Comment (randombit) [inline src/lib/stream/ctr/ctr.cpp]: CTR should throw Not_Implemented("CTR_BE::seek") - CTR supports this but it does not have to be added in this commit.

**Comment by randombit** [comment]
Your new changes and the additional tests look good. With the issue about throwing exceptions in the other implementations of `seek` that I just commented on addressed, this would be ready to go.

**Comment by SimCog** [comment]
Is this ok for you now?

**Comment by randombit** [comment]
This looks good. One final thing - can you rebase against master and squash these 13 commits down to a single one whose changelog has an accurate description of the final change (eg, adding StreamCipher::seek interface, supporting seek in ChaCha, and also adding ChaCha8 support).

If you haven't done it before, rebasing and squashing is pretty simple, though with the typical number of git footguns. These short docs explain everything you'll need to do:

https://help.github.com/articles/about-git-rebase/
https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

After rebasing you'll need to force push to update the PR.

**Comment by SimCog** [comment]
As asked, I squash my commits.

File: src/tests/test_stream.cpp [hunk 9]
```diff
@@ -39,7 +40,12 @@ class Stream_Cipher_Tests : public Text_Based_Test
 
          for(auto&& provider: providers)
             {
-            std::unique_ptr<Botan::StreamCipher> cipher(Botan::StreamCipher::create(algo, provider));
+            std::unique_ptr<Botan::StreamCipher> cipher;
+            try {
+               cipher = Botan::StreamCipher::create(algo, provider);
```
Comment (randombit) [inline src/tests/test_stream.cpp]: What's the deal with these changes to the tests? From my earlier review I didn't notice any changes to this file besides adding the Seek param.

Comment (SimCog) [inline src/tests/test_stream.cpp]: For the coding coverage. The tests didn't pass through the throw.

Comment (randombit) [inline src/tests/test_stream.cpp]: Remove please, just leaving the changes to adding the Seek param. Other than that this looks ready to merge.

**Comment by SimCog** [comment]
As asked, I removed some tests. I also removed named input parameters for the seek methods of openssl_rc4, rc4, ofb, ctr and salsa since there are warnings with it.

**Comment by randombit** [comment]
Thanks for doing this work, now merged for upcoming 1.11.30 release
